p2p/tools/network.py

```diff
@@ -0,0 +1,144 @@
+import asyncio
+from asyncio.base_events import Server
+import logging
+from typing import (
+    Dict,
+    Iterable,
+    NamedTuple,
+    Set,
+    Tuple,
+)
+
+
+class Address(NamedTuple):
+    transport: str
+    ip: str
+    port: int
+
+
+logger = logging.getLogger('p2p.testing.network')
+
+
+class MockServer(Server):
+    """
+    Mock `asyncio.Server` object.
+    """
+    def __init__(self, client_connected_cb, address, network):
+        self.client_connected_cb = client_connected_cb
+        self.address = address
+        self.network = network
+
+    def __repr__(self):
+        return '<%s %s:%s:%s>' % (self.__class__.__name__, self.address.transport, self.address.ip, self.address.port)
+
+    def close(self):
+        pass
+
+    async def wait_closed(self):
+        return
+
+
+class _MemoryTransport(asyncio.Transport):
+    """Direct connection between a StreamWriter and StreamReader."""
+
+    def __init__(self, reader: asyncio.StreamReader) -> None:
+        super().__init__()
+        self._reader = reader
+
+    def write(self, data: bytes) -> None:
+        self._reader.feed_data(data)
+
+    def writelines(self, data: Iterable[bytes]) -> None:
+        for line in data:
+            self._reader.feed_data(line)
+            self._reader.feed_data(b'\n')
+
+    def write_eof(self) -> None:
+        self._reader.feed_eof()
+
+    def can_write_eof(self) -> bool:
+        return True
+
+    def is_closing(self) -> bool:
+        return False
+
+    def close(self) -> None:
+        self.write_eof()
+
+
+def mempipe(ip, port) -> Tuple[asyncio.StreamReader, asyncio.StreamWriter]:
+    """In-memory pipe, returns a ``(reader, writer)`` pair.
+
+    .. versionadded:: 0.1
+
+    """
+
+    reader = asyncio.StreamReader()
+    writer = asyncio.StreamWriter(
+        transport=_MemoryTransport(reader),
+        protocol=asyncio.StreamReaderProtocol(reader),
+        reader=reader,
+        loop=asyncio.get_event_loop(),
+    )
+    return reader, writer
+
+
+def get_connected_readers(server, client):
+    server_reader, client_writer = mempipe()
+    client_reader, server_writer = mempipe()
+
+    server_reader.__id = 'lr'
+    server_writer.__id = 'lw'
+    client_reader.__id = 'rr'
+    client_writer.__id = 'rw'
+
+    return (
+        server_reader, server_writer,
+        client_reader, client_writer,
+    )
+
+
+# TODO: Need a `Router` and a `Network`.  Network is aware of IP address and Router
```

Comment:
This is needed so that `writer.get_extra_info('peername')` can return the proper data.  That means that we instantiate a `Network` object for each `IP` address (and this `Network` object is what gets passed to the `Server` and `PeerPool`.  The `Router` object is more of a global and handles managing connections between `Network` objects.